libcontainer/mount_linux.go

```diff
@@ -121,3 +179,75 @@ func syscallMode(i fs.FileMode) (o uint32) {
 	// No mapping for Go's ModeTemporary (plan9 only).
 	return
 }
+
+// mountFd creates a "mount source fd" (either through open_tree(2) or just
+// open(O_PATH)) based on the provided configuration. This function must be
+// called from within the container's mount namespace.
+//
+// In the case of idmapped mount configurations, the returned mount source will
+// be an open_tree(2) file with MOUNT_ATTR_IDMAP applied. For other
+// bind-mounts, it will be an O_PATH. If the type of mount cannot be handled,
+// the returned mountSource will be nil, indicating that the container init
+// process will need to do an old-fashioned mount(2) themselves.
+//
+// This helper is only intended to be used by goCreateMountSources.
+func mountFd(nsHandles *userns.Handles, m *configs.Mount) (*mountSource, error) {
+	if !m.IsBind() {
+		return nil, errors.New("new mount api: only bind-mounts are supported")
+	}
+	if nsHandles == nil {
+		nsHandles = new(userns.Handles)
+		defer nsHandles.Release()
+	}
+
+	var mountFile *os.File
+	var sourceType mountSourceType
+
+	if m.IsBind() {
+		// We only need to use OPEN_TREE_CLONE in the case where we need to use
+		// mount_setattr(2). We are currently in the container namespace and
+		// thus there is no risk of an opened directory being used to escape
+		// the container. OPEN_TREE_CLONE is more expensive than open(2)
+		// because it requires doing mounts inside a new anonymous mount
+		// namespace.
+		if m.IsIDMapped() {
+			flags := uint(unix.OPEN_TREE_CLONE | unix.O_CLOEXEC)
+			if m.Flags&unix.MS_REC == unix.MS_REC {
+				flags |= unix.AT_RECURSIVE
+			}
+			fd, err := unix.OpenTree(unix.AT_FDCWD, m.Source, flags)
+			if err != nil {
+				return nil, &os.PathError{Op: "open_tree(OPEN_TREE_CLONE)", Path: m.Source, Err: err}
+			}
+			mountFile = os.NewFile(uintptr(fd), m.Source)
+			sourceType = mountSourceOpenTree
+
+			// Configure the id mapping.
+			usernsFile, err := nsHandles.Get(userns.Mapping{
+				UIDMappings: m.UIDMappings,
+				GIDMappings: m.GIDMappings,
+			})
+			if err != nil {
+				return nil, fmt.Errorf("failed to create userns for %s id-mapping: %w", m.Source, err)
+			}
+			defer usernsFile.Close()
+			if err := unix.MountSetattr(int(mountFile.Fd()), "", unix.AT_EMPTY_PATH, &unix.MountAttr{
```

Comment:
I think we should also set `AT_RECURSIVE` here if we use `rbind`?
So, maybe we should add a testcase for `rbind` with `idmapped mount`?

Reply:
We haven't added support for `rbind` yet to runc (it still isn't in a released spec). I will do that in a future PR.

Reply:
I got it now. But you have checked it when calling `unix.OpenTree`. Do you think we should remove this `AT_RECURSIVE` and add a warning log when calling `unix.OpenTree`?
https://github.com/opencontainers/runc/blob/ae3fd6b345cb6e7de7b0373211af3ddbc7862166/libcontainer/mount_linux.go#L215-L217

Reply:
Yeah, I'll remove that -- I added it because we didn't have `rbind` at the time so it made the most sense to me to make `rbind` recursive.

Reply:
Actually, I think it makes more sense to add `idmap` and `ridmap` support.

Reply:
@lifubang That code is needed for `rbind` mounts to work. At the moment the implementation _does not_ recursively apply the id mapping configuration (`mount_setattr` is called without `AT_RECURSIVE`).

However, it's probably a better idea to just implement `ridmap` and `idmap` support.

Reply:
Yes. But I’m curious about how to deal with ‘rbind,idmap’. Maybe we should treat it the same as ‘bind,idmap’?

Reply:
`rbind` + `idmap` would mean you do an `rbind` but with only the top-level mount having the idmapping applied. `MS_BIND|MS_REC` is completely separate to `MOUNT_ATTR_IDMAP`. I already have patches adding support, I'll push them in a bit once I finish writing the tests.